Weird stacktrace when using an invalid Security Manager

Java -version output
--------------------
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-16.0.1+9 (build 16.0.1+9)
Eclipse OpenJ9 VM AdoptOpenJDK-16.0.1+9 (build openj9-0.26.0, JRE 16 Linux amd64-64-Bit Compressed References 20210421_23 (JIT enabled, AOT enabled)
OpenJ9   - b4cc246d9
OMR      - 162e6f729
JCL      - cea22090ecf based on jdk-16.0.1+9)

openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment (build 16.0.1+9-24)
OpenJDK 64-Bit Server VM (build 16.0.1+9-24, mixed mode, sharing)

Summary of problem
------------------
While playing around with the security manager, I've found out this weird stack trace: It includes the module name and version using slashes and the version is 9, even though I'm using JDK 16.

Diagnostic files
----------------
```
/jvms/jdk-16-openj9/bin/java -Djava.security.manager=x -version
Exception in thread "(unnamed thread)" java/lang/Error: JVM can't set custom SecurityManager due to java.lang.ClassNotFoundException: x
        at java/lang/ClassLoader.initializeClassLoaders (java.base@9/ClassLoader.java:240)
        at java/lang/Thread.initialize (java.base@9/Thread.java:444)
        at java/lang/Thread.<init> (java.base@9/Thread.java:161)
```
This is the output of Hotspot:
```
/jvms/jdk-16/bin/java -Djava.security.manager=x -version
Error occurred during initialization of VM
java.lang.InternalError: Could not create SecurityManager
        at java.lang.System.initPhase3(java.base@16.0.1/System.java:2120)
Caused by: java.lang.ClassNotFoundException: x
        at jdk.internal.loader.BuiltinClassLoader.loadClass(java.base@16.0.1/BuiltinClassLoader.java:636)
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(java.base@16.0.1/ClassLoaders.java:182)
        at java.lang.ClassLoader.loadClass(java.base@16.0.1/ClassLoader.java:519)
        at java.lang.Class.forName0(java.base@16.0.1/Native Method)
        at java.lang.Class.forName(java.base@16.0.1/Class.java:466)
        at java.lang.System.initPhase3(java.base@16.0.1/System.java:2105)
```

--- a/runtime/ddr/j9ddr.h
+++ b/runtime/ddr/j9ddr.h
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2014 IBM Corp. and others
+ * Copyright (c) 1991, 2021 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -26,11 +26,6 @@
 
 #ifdef __cplusplus
 extern "C" {
-#endif
-
-#ifndef J9_STR
-#define J9_STR_(x) #x
-#define J9_STR(x) J9_STR_(x)
 #endif
 
 struct OMRPortLibrary;

--- a/runtime/jcl/common/jclexception.c
+++ b/runtime/jcl/common/jclexception.c
@@ -110,7 +110,7 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 	/* Create the new StackTraceElement and put it in the array at the correct index */
 
 	element = mmfns->J9AllocateObject(vmThread, userData->elementClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
-	if (element == NULL) {
+	if (NULL == element) {
 		rc = FALSE;
 		vmFuncs->setHeapOutOfMemoryError(vmThread);
 	} else {
@@ -119,12 +119,10 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 		userData->index += 1;
 
 		/* If there is a valid method at this frame, fill in the information for it in the StackTraceElement */
-
-		if (romMethod != NULL) {
+		if (NULL != romMethod) {
 			J9UTF8 const * utfClassName = J9ROMCLASS_CLASSNAME(romClass);
 			J9UTF8 * utf = NULL;
 			j9object_t string = NULL;
-			UDATA j2seVersion = J2SE_VERSION(vm) & J2SE_VERSION_MASK;
 
 			PUSH_OBJECT_IN_SPECIAL_FRAME(vmThread, element);
 
@@ -146,15 +144,14 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 			}
 
 			/* Fill in module name and version */
-			if (j2seVersion >= J2SE_V11) {
-				J9Module *module = NULL;
-				if (classLoader != NULL) {
+			if (JAVA_SPEC_VERSION >= 11) {
+				if (NULL != classLoader) {
 					j9object_t classLoaderName = J9VMJAVALANGCLASSLOADER_CLASSLOADERNAME(vmThread, classLoader->classLoaderObject);
 					J9VMJAVALANGSTACKTRACEELEMENT_SET_CLASSLOADERNAME(vmThread, element, classLoaderName);
 				}
 				if (J9_ARE_NO_BITS_SET(vm->runtimeFlags, J9_RUNTIME_JAVA_BASE_MODULE_CREATED)) {
 					string = mmfns->j9gc_createJavaLangString(vmThread, (U_8 *)JAVA_BASE_MODULE, LITERAL_STRLEN(JAVA_BASE_MODULE), J9_STR_XLAT);
-					if (string == NULL) {
+					if (NULL == string) {
 						rc = FALSE;
 						/* exception is pending from the call */
 						goto done;
@@ -162,8 +159,8 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 					element = PEEK_OBJECT_IN_SPECIAL_FRAME(vmThread, 0);
 					J9VMJAVALANGSTACKTRACEELEMENT_SET_MODULENAME(vmThread, element, string);
 
-					string = mmfns->j9gc_createJavaLangString(vmThread, (U_8 *)JAVA_MODULE_VERSION, LITERAL_STRLEN(JAVA_MODULE_VERSION), J9_STR_XLAT);
-					if (string == NULL) {
+					string = mmfns->j9gc_createJavaLangString(vmThread, (U_8 *)JAVA_SPEC_VERSION_STRING, LITERAL_STRLEN(JAVA_SPEC_VERSION_STRING), J9_STR_XLAT);
+					if (NULL == string) {
 						rc = FALSE;
 						/* exception is pending from the call */
 						goto done;
@@ -174,7 +171,8 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 					/* Fetch the J9Module from the j.l.Class->j.l.Module field if we have a class.
 					 * Otherwise the more painful package-based lookup must be performed
 					 */
-					if (ramClass != NULL) {
+					J9Module *module = NULL;
+					if (NULL != ramClass) {
 						j9object_t moduleObject = J9VMJAVALANGCLASS_MODULE(vmThread, ramClass->classObject);
 						module = J9OBJECT_ADDRESS_LOAD(vmThread, moduleObject, vm->modulePointerOffset);
 					} else {
@@ -183,7 +181,6 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 						module = vmFuncs->findModuleForPackage(vmThread, classLoader, J9UTF8_DATA(utfClassName), (U_32) length);
 						omrthread_monitor_exit(vm->classLoaderModuleAndLocationMutex);
 					}
-
 					if (NULL != module) {
 						J9VMJAVALANGSTACKTRACEELEMENT_SET_MODULENAME(vmThread, element, module->moduleName);
 						J9VMJAVALANGSTACKTRACEELEMENT_SET_MODULEVERSION(vmThread, element, module->version);
@@ -210,11 +207,9 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 						rc = FALSE;
 						/* exception is pending from the call */
 						goto done;
-					} else {
-						if (NULL != ramClass) {
-							/* Class name was interned so it's safe to write it back to the Class Object */
-							J9VMJAVALANGCLASS_SET_CLASSNAMESTRING(vmThread, J9VM_J9CLASS_TO_HEAPCLASS(ramClass), string);
-						}
+					} else if (NULL != ramClass) {
+						/* Class name was interned so it's safe to write it back to the Class Object */
+						J9VMJAVALANGCLASS_SET_CLASSNAMESTRING(vmThread, J9VM_J9CLASS_TO_HEAPCLASS(ramClass), string);
 					}
 				}
 				element = PEEK_OBJECT_IN_SPECIAL_FRAME(vmThread, 0);
@@ -224,7 +219,7 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 			/* Fill in method name - intern the string as it's also interned by the StackWalker API and by Reflection */
 			utf = J9ROMMETHOD_NAME(romMethod);
 			string = mmfns->j9gc_createJavaLangString(vmThread, J9UTF8_DATA(utf), (U_32) J9UTF8_LENGTH(utf), J9_STR_TENURE | J9_STR_INTERN);
-			if (string == NULL) {
+			if (NULL == string) {
 				rc = FALSE;
 				/* exception is pending from the call */
 				goto done;
@@ -243,36 +238,34 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 			 * This avoids additional allocations during stack trace generation
 			 */
 			string = NULL;
-			if (ramClass != NULL) {
+			if (NULL != ramClass) {
 				string = J9VMJAVALANGCLASS_FILENAMESTRING(vmThread, J9VM_J9CLASS_TO_HEAPCLASS(ramClass));
 			}
-			if (string == NULL) {
-				if (fileName != NULL) {
-					J9UTF8 *previousFileName = userData->previousFileName;
-					/* Use an == comparison here as the previousFileName may have been from
-					* a classloader that was unloaded.  We can safely do an == comparison
-					* as we know the current class is deeper in the stack and can't have
-					* incorrectly been loaded into the space the previous loader was removed
-					* from.  We can't do a string compare here but the == should be sufficient
-					* provided the utf8 interning is hitting on common strings.
-					*/
-					if (previousFileName == fileName) {
-						j9array_t result = (j9array_t) PEEK_OBJECT_IN_SPECIAL_FRAME(vmThread, 2);
-						element = J9JAVAARRAYOFOBJECT_LOAD(vmThread, result, currentIndex - 1);
-						string = J9VMJAVALANGSTACKTRACEELEMENT_FILENAME(vmThread, element);
-					} else {
-						string = mmfns->j9gc_createJavaLangString(vmThread, J9UTF8_DATA(fileName), (U_32) J9UTF8_LENGTH(fileName), 0);
-						if (string == NULL) {
-							rc = FALSE;
-							/* exception is pending from the call */
-							goto done;
-						}
+			if ((NULL == string) && (NULL != fileName)) {
+				J9UTF8 *previousFileName = userData->previousFileName;
+				/* Use an == comparison here as the previousFileName may have been from
+				* a classloader that was unloaded.  We can safely do an == comparison
+				* as we know the current class is deeper in the stack and can't have
+				* incorrectly been loaded into the space the previous loader was removed
+				* from.  We can't do a string compare here but the == should be sufficient
+				* provided the utf8 interning is hitting on common strings.
+				*/
+				if (previousFileName == fileName) {
+					j9array_t result = (j9array_t) PEEK_OBJECT_IN_SPECIAL_FRAME(vmThread, 2);
+					element = J9JAVAARRAYOFOBJECT_LOAD(vmThread, result, currentIndex - 1);
+					string = J9VMJAVALANGSTACKTRACEELEMENT_FILENAME(vmThread, element);
+				} else {
+					string = mmfns->j9gc_createJavaLangString(vmThread, J9UTF8_DATA(fileName), (U_32) J9UTF8_LENGTH(fileName), 0);
+					if (NULL == string) {
+						rc = FALSE;
+						/* exception is pending from the call */
+						goto done;
 					}
-					Assert_JCL_notNull(string);
-					if (ramClass != NULL) {
-						/* Update the cached fileNameString on the class so subsequent calls will find it */
-						J9VMJAVALANGCLASS_SET_FILENAMESTRING(vmThread, J9VM_J9CLASS_TO_HEAPCLASS(ramClass), string);
-					}
+				}
+				Assert_JCL_notNull(string);
+				if (NULL != ramClass) {
+					/* Update the cached fileNameString on the class so subsequent calls will find it */
+					J9VMJAVALANGCLASS_SET_FILENAMESTRING(vmThread, J9VM_J9CLASS_TO_HEAPCLASS(ramClass), string);
 				}
 			}
 			/* Update previous filename as it must always match the contents of the StackTraceElement[n-1]'s
@@ -280,21 +273,20 @@ getStackTraceIterator(J9VMThread * vmThread, void * voidUserData, UDATA bytecode
 			 * name into the StackTraceElement
 			 */
 			userData->previousFileName = fileName;
-			if (string != NULL) {
+			if (NULL != string) {
 				element = PEEK_OBJECT_IN_SPECIAL_FRAME(vmThread, 0);
 				J9VMJAVALANGSTACKTRACEELEMENT_SET_FILENAME(vmThread, element, string);
 			}
 
 			/* Fill in line number - Java wants -2 for natives, -1 for no line number (which will be 0 coming in from the iterator) */
-
-			if (romMethod->modifiers & J9AccNative) {
+			if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccNative)) {
 				lineNumber = -2;
-			} else if (lineNumber == 0) {
+			} else if (0 == lineNumber) {
 				lineNumber = -1;
 			}
 			J9VMJAVALANGSTACKTRACEELEMENT_SET_LINENUMBER(vmThread, element, (I_32) lineNumber);
 
-			if (vm->verboseLevel & VERBOSE_STACKTRACE) {
+			if (J9_ARE_ANY_BITS_SET(vm->verboseLevel, VERBOSE_STACKTRACE)) {
 				setStackTraceElementSource(vmThread, element, classLoader, romClass);
 			}
 

--- a/runtime/jvmti/jvmtiCapability.c
+++ b/runtime/jvmti/jvmtiCapability.c
@@ -41,8 +41,6 @@ dumpCapabilities(J9JavaVM * vm, const jvmtiCapabilities *capabilities, const cha
 
 	j9tty_printf(PORTLIB, "%s\n", caption);
 
-#define J9_STR_(x) #x
-#define J9_STR(x) J9_STR_(x)
 #define PRINT_CAPABILITY(capability) if (capabilities->capability) j9tty_printf(PORTLIB, "\t%s\n", J9_STR(capability));
 
 	PRINT_CAPABILITY(can_tag_objects);

--- a/runtime/oti/j9comp.h
+++ b/runtime/oti/j9comp.h
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2018 IBM Corp. and others
+ * Copyright (c) 1991, 2021 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -29,6 +29,9 @@
 #define J9_ARE_ANY_BITS_SET(value, bits) OMR_ARE_ANY_BITS_SET((value), (bits))
 #define J9_ARE_ALL_BITS_SET(value, bits) OMR_ARE_ALL_BITS_SET((value), (bits))
 #define J9_ARE_NO_BITS_SET(value, bits) OMR_ARE_NO_BITS_SET((value), (bits))
+
+#define J9_STR_(x) #x
+#define J9_STR(x) J9_STR_(x)
 
 #if defined(RS6000) || defined(LINUX) || defined(OSX)
 #define J9UNIX

--- a/runtime/oti/jvminit.h
+++ b/runtime/oti/jvminit.h
@@ -616,7 +616,6 @@ enum INIT_STAGE {
 #define SYSPROP_JDK_MODULE_PATCH "jdk.module.patch."
 #define SYSPROP_JDK_MODULE_ILLEGALACCESS "jdk.module.illegalAccess"
 #define JAVA_BASE_MODULE "java.base"
-#define JAVA_MODULE_VERSION "9"
 
 #define SYSPROP_COM_SUN_MANAGEMENT "-Dcom.sun.management."
 

--- a/runtime/vm/exceptiondescribe.c
+++ b/runtime/vm/exceptiondescribe.c
@@ -101,15 +101,15 @@ printExceptionMessage(J9VMThread* vmThread, j9object_t exception) {
 /* assumes VM access */
 static UDATA
 printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeOffset, J9ROMClass *romClass, J9ROMMethod * romMethod, J9UTF8 * sourceFile, UDATA lineNumber, J9ClassLoader* classLoader, J9Class* ramClass) {
-	const char* format;
+	const char* format = NULL;
 	J9JavaVM *vm = vmThread->javaVM;
 	J9InternalVMFunctions * vmFuncs = vm->internalVMFunctions;
 	PORT_ACCESS_FROM_JAVAVM(vm);
 
-	if (romMethod == NULL) {
+	if (NULL == romMethod) {
 		format = j9nls_lookup_message(
-			J9NLS_INFO | J9NLS_DO_NOT_PRINT_MESSAGE_TAG, 
-			J9NLS_VM_STACK_TRACE_UNKNOWN, 
+			J9NLS_INFO | J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
+			J9NLS_VM_STACK_TRACE_UNKNOWN,
 			NULL);
 		j9tty_err_printf(PORTLIB, (char*)format);
 	} else {
@@ -123,12 +123,8 @@ printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeO
 		char versionBuf[J9VM_PACKAGE_NAME_BUFFER_LENGTH];
 		BOOLEAN freeModuleName = FALSE;
 		BOOLEAN freeModuleVersion = FALSE;
-		UDATA j2seVersion = J2SE_VERSION(vm) & J2SE_VERSION_MASK;
 
-		if (j2seVersion >= J2SE_V11) {
-			moduleNameUTF = JAVA_BASE_MODULE;
-			moduleVersionUTF = JAVA_MODULE_VERSION;
-
+		if (JAVA_SPEC_VERSION >= 11) {
 			if (J9_ARE_ALL_BITS_SET(vm->runtimeFlags, J9_RUNTIME_JAVA_BASE_MODULE_CREATED)) {
 				J9Module *module = NULL;
 				U_8 *classNameUTF = J9UTF8_DATA(J9ROMCLASS_CLASSNAME(romClass));
@@ -138,26 +134,37 @@ printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeO
 				module = vmFuncs->findModuleForPackage(vmThread, classLoader, classNameUTF, (U_32) length);
 				omrthread_monitor_exit(vm->classLoaderModuleAndLocationMutex);
 
-				if ((NULL != module) && (module != vm->javaBaseModule)) {
-					moduleNameUTF = copyStringToUTF8WithMemAlloc(
-						vmThread, module->moduleName, J9_STR_NULL_TERMINATE_RESULT, "", 0, nameBuf, J9VM_PACKAGE_NAME_BUFFER_LENGTH, NULL);
-					if (nameBuf != moduleNameUTF) {
-						freeModuleName = TRUE;
+				if (NULL != module) {
+					if (module != vm->javaBaseModule) {
+						moduleNameUTF = copyStringToUTF8WithMemAlloc(
+							vmThread, module->moduleName, J9_STR_NULL_TERMINATE_RESULT, "", 0, nameBuf, J9VM_PACKAGE_NAME_BUFFER_LENGTH, NULL);
+						if (nameBuf != moduleNameUTF) {
+							freeModuleName = TRUE;
+						}
+					} else {
+						moduleNameUTF = JAVA_BASE_MODULE;
 					}
+
 					if (NULL != moduleNameUTF) {
 						moduleVersionUTF = copyStringToUTF8WithMemAlloc(
 							vmThread, module->version, J9_STR_NULL_TERMINATE_RESULT, "", 0, versionBuf, J9VM_PACKAGE_NAME_BUFFER_LENGTH, NULL);
-						if (versionBuf != moduleVersionUTF) {
+						if (NULL == moduleVersionUTF) {
+							moduleVersionUTF = JAVA_SPEC_VERSION_STRING;
+						} else if (versionBuf != moduleVersionUTF) {
 							freeModuleVersion = TRUE;
 						}
 					}
 				}
+			} else {
+				/* Assume bootstrap code within java.base when the module itself hasn't been created yet  */
+				moduleNameUTF = JAVA_BASE_MODULE;
+				moduleVersionUTF = JAVA_SPEC_VERSION_STRING;
 			}
 		}
-		if (romMethod->modifiers & J9AccNative) {
+		if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccNative)) {
 			sourceFileName = "NativeMethod";
 			sourceFileNameLen = LITERAL_STRLEN("NativeMethod");
-		} else if (sourceFile) {
+		} else if (NULL != sourceFile) {
 			sourceFileName = (char *)J9UTF8_DATA(sourceFile);
 			sourceFileNameLen = (UDATA)J9UTF8_LENGTH(sourceFile);
 		} else {
@@ -165,51 +172,31 @@ printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeO
 			sourceFileNameLen = LITERAL_STRLEN("Unknown Source");
 		}
 		if (NULL != moduleNameUTF) {
-			if (NULL != moduleVersionUTF) {
-				if (lineNumber) {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION_LINE,
-						"\tat %.*s.%.*s (%s@%s/%.*s:%u)\n");
-				} else {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION,
-						"\tat %.*s.%.*s (%s@%s/%.*s)\n");
-				}
-				j9tty_err_printf(PORTLIB, (char*)format,
-					(UDATA)J9UTF8_LENGTH(className), J9UTF8_DATA(className),
-					(UDATA)J9UTF8_LENGTH(methodName), J9UTF8_DATA(methodName),
-					moduleNameUTF, moduleVersionUTF,
-					sourceFileNameLen, sourceFileName,
-					lineNumber); /* line number will be ignored in if it's not used in the format string */
-				if (TRUE == freeModuleVersion) {
-					j9mem_free_memory(moduleVersionUTF);
-				}
+			if (0 != lineNumber) {
+				format = j9nls_lookup_message(
+					J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
+					J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION_LINE,
+					"\tat %.*s.%.*s (%s@%s/%.*s:%u)\n");
 			} else {
-				if (lineNumber) {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE_LINE,
-						"\tat %.*s.%.*s (%s/%.*s:%u)\n");
-				} else {
-					format = j9nls_lookup_message(
-						J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
-						J9NLS_VM_STACK_TRACE_WITH_MODULE,
-						"\tat %.*s.%.*s (%s/%.*s)\n");
-				}
-				j9tty_err_printf(PORTLIB, (char*)format,
-					(UDATA)J9UTF8_LENGTH(className), J9UTF8_DATA(className),
-					(UDATA)J9UTF8_LENGTH(methodName), J9UTF8_DATA(methodName),
-					moduleNameUTF,
-					sourceFileNameLen, sourceFileName,
-					lineNumber); /* line number will be ignored in if it's not used in the format string */
+				format = j9nls_lookup_message(
+					J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
+					J9NLS_VM_STACK_TRACE_WITH_MODULE_VERSION,
+					"\tat %.*s.%.*s (%s@%s/%.*s)\n");
+			}
+			j9tty_err_printf(PORTLIB, (char*)format,
+				(UDATA)J9UTF8_LENGTH(className), J9UTF8_DATA(className),
+				(UDATA)J9UTF8_LENGTH(methodName), J9UTF8_DATA(methodName),
+				moduleNameUTF, moduleVersionUTF,
+				sourceFileNameLen, sourceFileName,
+				lineNumber); /* line number will be ignored in if it's not used in the format string */
+			if (TRUE == freeModuleVersion) {
+				j9mem_free_memory(moduleVersionUTF);
 			}
 			if (TRUE == freeModuleName) {
 				j9mem_free_memory(moduleNameUTF);
 			}
 		} else {
-			if (lineNumber) {
+			if (0 != lineNumber) {
 				format = j9nls_lookup_message(
 					J9NLS_DO_NOT_PRINT_MESSAGE_TAG,
 					J9NLS_VM_STACK_TRACE_WITH_LINE,

--- a/runtime/vm/vmprops.c
+++ b/runtime/vm/vmprops.c
@@ -575,9 +575,7 @@ initializeSystemProperties(J9JavaVM * vm)
 	if (JAVA_SPEC_VERSION == 8) {
 		specificationVersion = "1.8";
 	} else {
-#define J9_STR_(x) #x
-#define J9_STR(x) J9_STR_(x)
-		specificationVersion = J9_STR(JAVA_SPEC_VERSION);
+		specificationVersion = JAVA_SPEC_VERSION_STRING;
 	}
 
 	/* Some properties (*.vm.*) are owned by the VM and need to be set early for all

